Refactor reusable block edit component using hooks (and fix interactions with multiple instances of the same reusable block)

ZebulanStanphill opened the pull request:
## Description
This PR refactors the `ReusableBlockEdit` to use React hooks.
It also fixes two issues with the current implementation in `master`:
- Editing/saving an instance of a reusable block didn't automatically update other instances in the editor.
- Editing a reusable block, putting focus in the title field, and then pressing `Esc` to cancel editing the reusable block didn't work properly. (It would mess up the preview of the reusable block in the editor.)

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
+	const { ref } = attributes;
+
+	const {
+		theReusableBlock,
```
I see the problem you're addressing here. Can't say I love the "the" prefix as a generalized solution. I guess it'll continue to be an issue we deal with, save for some other prior attempts to work around this (#21057, [#18973](https://github.com/WordPress/gutenberg/pull/18973/files#diff-2a00f67990de843497e20b203dcf4084R23)). The other option I might resort to is something "less ideal" inside the selector, like an underscore prefix `_`, so at least the desirable name `reusableBlock` is granted for the primary usage scope.

ZebulanStanphill replied:
The underscore pattern seems pretty reasonable to me; I'll use that.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
+		/** @typedef {import('@wordpress/blocks').BlockInstance} BlockInstance */
+
+		/**
+		 * @typedef WPReusableBlockEditSelectedData
```
I guess there's value in having it located near where it's used, but I'd personally prefer most custom type definitions to be consolidated toward the top of a file, so as to be a consistent reference point. It also helps avoid the issue you've run into with line length on the property descriptions.

aduth replied:
Also, this doesn't feel like something we should have to create a type for. If the selector return value is properly typed, then TypeScript should be able to infer the return value for us, which would go a long way to avoid human error and duplicate typing.

I kinda feel all this typing should be removed. I don't think we can or should expect developers to add types like this in every usage of `useSelect`, and if we do, it's a failing on the part of the framework tooling.

ZebulanStanphill replied:
Currently, the types aren't inferred because the functions returned from the `select` calls don't have type definitions. Now that you point it out, the real solution would be to add types to those where they're defined. I'll remove the type definition from this PR.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
+		settings,
+		title,
+	} = useSelect(
+		/** @typedef {import('@wordpress/blocks').BlockInstance} BlockInstance */
```
If you really want this file to be type-checked, you could consider the approach I took in #21362 with staring to progressively opt-in a package.

ZebulanStanphill replied:
There's not much to type here until the `select` functions have type definitions, so there's not point in turning on type-checking for this file, currently.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
-	componentDidMount() {
-		if ( ! this.props.reusableBlock ) {
-			this.props.fetchReusableBlock();
+	}, EMPTY_ARRAY );
```
Can you justify this? It seems on the end of being a premature optimization.

```suggestion
	}, [] );
```

ZebulanStanphill replied:
Good point. I'll avoid the constant for now.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,188 @@ import { partial } from 'lodash';
+		 *                                                    block being fetched.
+		 * @property {boolean}               isSaving         Is the reusable
+		 *                                                    block being saved.
+		 * @property {?boolean}              isTemporary      Is the reusable
+		 *                                                    block temporary.
```
To my previous point, it's also not obvious if `isTemporary` should actually be nullable, or if this was just an oversight in the initial implementation. A consistent boolean type seems preferable, which we could change in the reducer / selector / wherever this is produced. To need to make that change and then return here to update these types is going to cause a lot of syncing errors / inconsistencies.

ZebulanStanphill replied:
Sometimes the reusable block is `null`, so it didn't seem right to always return a boolean for a value which is derived from an object that may or may not exist.

ZebulanStanphill replied:
To clarify, `isTemporary` can be `null`. I'm getting it using `_reusableBlock?.title ?? null`. I could default it to `false`, but I'm not sure if that's technically the right thing to do.

aduth replied:
My point here was more about: It may be true that the selector returns a `null` value, but it may very well be _wrong_ in doing so. We can fix that at the selector, but if we had kept these duplicate types here, it would be at high risk for falling out of sync.

ZebulanStanphill replied:
Now that the types are gone, do you think is fine for now?

aduth replied:
> Now that the types are gone, do you think this is fine for now?

Yes, the issue was mostly about the type definitions.

aduth reviewed (commented):
Seems like a good candidate for a refactoring. Reusable blocks in general could do for a rethink (also from how it's stored in state. This could help prepare for this and avoid some inevitable future changes.

ZebulanStanphill commented:
>Reusable blocks in general could do for a rethink (also from how it's stored in state. This could help prepare for this and avoid some inevitable future changes.

Yeah, I was thinking about trying to make some bigger changes to Reusable Blocks, but I had to improve their implementation first.

Except possibly the `isTemporary` thing, I've addressed all of your feedback, @aduth. :slightly_smiling_face: I've removed all type definitions from this file, since a lot of the typing should actually happen elsewhere.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,158 @@ import { partial } from 'lodash';
+				isFetching: isFetchingReusableBlock( ref ),
+				isSaving: isSavingReusableBlock( ref ),
+				isTemporary: _reusableBlock?.isTemporary ?? null,
+				blocks: _reusableBlock ? parse( _reusableBlock.content ) : null,
```
`parse` here is concerning, since the `useSelect` selector will be called very frequently for any reusable block which is mounted in the page (every state change for every store). The selector should be as efficient as possible, and `parse` isn't an especially trivial operation. In the previous implementation, we apparently relied on memoization of `__experimentalGetParsedReusableBlock` to avoid calling `parse` more than necessary. Is there any reason that was dropped here?

ZebulanStanphill replied:
It was an oversight on my part. Thanks for catching! I've restored the usage of `__experimentalGetParsedReusableBlock`.

ZebulanStanphill replied:
It turns out that trying to use `__experimentalGetParsedReusableBlock( _reusableBlock.id )` instead of `parse( _reusableBlock.content )` was causing end-to-end test failures and broken behavior in my testing. I'm not sure why, but for now, I've reverted back to using the `parse` approach.

aduth replied:
Ok, but that brings us back to the original concern in https://github.com/WordPress/gutenberg/pull/21427#discussion_r424666982 . 

Furthermore, we shouldn't be okay to accept a solution because we don't understand a failing of the preferred option. And `__experimentalGetParsedReusableBlock` must exist for a reason, or at least: We should want to use it or eliminate it consistently.

What errors were you seeing?

ZebulanStanphill replied:
I only reverted back to `parse` temporarily to focus on the [other issue](https://github.com/WordPress/gutenberg/pull/21427#discussion_r424667843) you brought up. I don't think we should merge this until we figure out this problem, too.

I'll push a commit redoing the switch back to `__experimentalGetParsedReusableBlock` to make it easier to test once I'm finished with the other issue.

(Testing stuff has taken a really long time ever since [`wp-env` stopped working on Linux](https://github.com/WordPress/gutenberg/issues/20180) and I've had to rely on `npm run package-plugin` or gutenberg.run.)

aduth replied:
Ah, fair enough, sorry for jumping to conclusions 😄

ZebulanStanphill replied:
Alright, I've redone the commit to switch to `__experimentalGetParsedReusableBlock`. It seems to completely break reusable block creation. I have no idea why. But anyway, it should be easy for others to test and help figure it out now.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,158 @@ import { partial } from 'lodash';
+	let content = (
+		<BlockEditorProvider
+			settings={ settings }
+			value={ localBlocks }
```
Personally, I might have implemented this slightly different to avoid storing two copies of the same blocks array, where `localBlocks` would only be set if editing, then this could be `value={ localBlocks ?? blocks }`

ZebulanStanphill replied:
That's a good idea. I'm currently working on trying it out locally. It might be a bit tricky to implement depending on how the `onChange` and `onInput` callbacks on `BlockEditorProvider` work, though. I need to make sure the local state remains `null`ed when not editing or else the bugs I fixed will probably become unfixed.

ZebulanStanphill replied:
I tried to do this in https://github.com/WordPress/gutenberg/pull/21427/commits/f16815a21d4926b4b074bb3e1a961dbc1ec7bab3, but it's causing reusable blocks to simply not load properly at all. Any ideas as to why?

aduth replied:
> I tried to do this in [f16815a](https://github.com/WordPress/gutenberg/commit/f16815a21d4926b4b074bb3e1a961dbc1ec7bab3), but it's causing reusable blocks to simply not load properly at all. Any ideas as to why?

In the commit, you're providing a function as `value` to `BlockEditorProvider` which, as far as I know, is not supported.

Shouldn't it be:

```js
value={ isEditing ? localBlocks : blocks }
```

ZebulanStanphill replied:
Thanks! Things are almost fixed now, but there's still one issue: saving a modified reusable block crashes the editor and outputs [React error 185](https://reactjs.org/docs/error-decoder.html/?invariant=185) in the JS console.

So it looks like there's some kind of recursion issue, but I can't figure out where.

ZebulanStanphill replied:
Alright, this is working now.

chrisvanpatten commented:
Not to let perfect be the enemy of good, but is there any value in holding until controlled InnerBlocks is stabilized, and refactor reusable blocks to leverage controlled InnerBlocks as part of these changes?

ZebulanStanphill commented:
@chrisvanpatten Since this PR fixes some annoying bugs, I'd say it's worth getting in sooner rather than later. The switch to hooks should make future changes easier anyway.

Also, I'm not entirely sure what controlled `InnerBlocks` even means. Could you direct me to where the discussion on this feature is happening?

ZebulanStanphill commented:
The current test failures are not related to this PR; it looks like the version bump to Gutenberg 8.1.0 in `master` was a bit premature:

```
[1] /home/travis/build/WordPress/gutenberg/packages/api-fetch/src/index.js
[1]   168:3  error  Deprecated functions must be removed before releasing this version  no-restricted-syntax
[1] 
[1] ✖ 1 problem (1 error, 0 warnings)
```

aduth commented:
> The current test failures are not related to this PR; it looks like the version bump to Gutenberg 8.1.0 in `master` was a bit premature:

See: #22333

aduth commented:
You'll probably have to rebase once more, to account for the merged changes of #22333.

ZebulanStanphill committed:
Refactor reusable block edit component using hooks.

ZebulanStanphill committed:
Only use local state while editing.

ZebulanStanphill committed:
Use __experimentalGetParsedReusableBlock.

ZebulanStanphill commented:
As per [this feedback](https://github.com/WordPress/gutenberg/pull/21427#discussion_r424666982), I've switched to using `__experimentalGetParsedReusableBlock` (similar to the approach used in `master`) in https://github.com/WordPress/gutenberg/pull/21427/commits/5f10743b7eb1105e79104af12cae6930648128f2. However, this is breaking a lot of stuff. Any feedback to help figure out why would be greatly appreciated.

aduth commented:
It is quite strange, and not immediately clear to me what the issue might be. It seems there's a bit of a delicate initialization of state based on whether the block is "temporary". I also wonder if moving from the higher-order component (`withSelect`, etc) to inline hooks might be changing something about the order in which `getReusableBlock` becomes available and/or has `isTemporary` set in a way we were anticipating previously. At least looking at the state transitions through Redux DevTools, it seems like the underlying data is being saved correctly. Switching between Code and back to visual will restore the broken block markup. It almost makes me wonder if the component isn't being updated in response to the state changes, though I don't know why that would ever be the case.

I'd have to spend some more time looking at it. I can try to find some time in my morning tomorrow.

aduth commented on packages/block-library/src/block/edit.js:
```diff
@@ -6,207 +6,182 @@ import { partial } from 'lodash';
-			this.state = {
-				isEditing: reusableBlock.isTemporary,
-				title: reusableBlock.title,
-				blocks: parse( reusableBlock.content ),
```
To the recent issues you're experiencing, I observe the difference between the branch and master being that, at the initialization of state for a new reusable block, we'd previously parse the content of the temporary block. In the new implementation, the blocks state will be initialized to `null` instead, and never be updated.

I think we need to ensure the state is initialized with the parsed result, and optionally have effects in place which keep that state up to date in response to changes (is there an equivalent to what existed previously in `componentDidUpdate`?).

Per the prior concern about performance, we'd still want to make sure that the content is parsed only once for the initial state.

See: https://reactjs.org/docs/hooks-reference.html#lazy-initial-state

ZebulanStanphill replied:
Thanks. Something I don't understand is how detecting the existence of `reusableBlock` indicates a _newly created_ reusable block. I guess I don't quite understand what that if statement means.

The equivalent to the `componentDidUpdate` code were [these effects](https://github.com/WordPress/gutenberg/pull/21427/commits/78767e8362e37e4f6f44872346ece6f2257eecec#diff-ae0ba49328ef77ebd8fc08a86fee257aR92-R102). But since you wanted to stop using local state when not necessary, I removed those.

aduth replied:
Hm, I think it may have been a pretty fragile assumption based on how an _existing_ reusable block would need to be fetched (see `fetchReusableBlock` in prior `componentDidMount`) and thus there was an assumption that `reusableBlock` would only be available initially in the case of a newly converted reusable block. It seems like something that would be more reliable to check for `reusableBlock?.isTemporary`.

ZebulanStanphill replied:
Alright, I'm pushing a commit right now that might fix things... I'll let you know how it turns out. Thank you very much for all of your help so far. :smile:

ZebulanStanphill replied:
Well, it didn't work. I think I probably missed something, though. This is all rather difficult to wrap my head around. :confounded:

ZebulanStanphill replied:
Alright, I messed around with the code some more, and now I've got things working again! :partying_face:

There's one minor issue I've noticed, however: whenever you switch between editing/not-editing a reusable block, it momentarily appears empty (IOW, it temporarily reverts to just showing the empty Paragraph placeholder). It makes for some annoying jumpiness that definitely isn't ideal.

ZebulanStanphill replied:
It turns out that the momentary emptiness was easily fixed by switching from `useEffect` to `useLayoutEffect`. :smile:

ZebulanStanphill committed:
Attempt to fix temporary reusable block behavior.

chrisvanpatten reviewed (approved):
This looks good to me. Are there any other reviews needed?

ZebulanStanphill commented:
Noting that @chrisvanpatten has [expressed interest](https://wordpress.slack.com/archives/C02QB2JS7/p1593008626189000?thread_ts=1593008189.179500&cid=C02QB2JS7) in creating a PR to replace this one, making use of the controlled `InnerBlocks` API. Until then, this PR remains stuck due to the aforementioned remaining regression in this branch.

ZebulanStanphill committed:
Fix mometary emptiness on start/stop editing.

ZebulanStanphill commented:
https://github.com/WordPress/gutenberg/pull/21427/commits/62852c55a2ab5d0826bc2553779e63b190db1a3a fixed the last issue with this PR. It should be good to go now. :partying_face:

ZebulanStanphill commented:
I'd like a final code review prior to merge. Since the bugs this PR fixes are rather significant (in my opinion), I've added the "Backport to WP Core" label. Feel free to remove it if you disagree.

ZebulanStanphill commented:
Rebased. @aduth, if you have the time, could you review this again? (Your previous review is blocking.)

aduth commented:
I don't know that I can review this right away, but I can at least dismiss my own older review 👍

ZebulanStanphill commented:
I tested this branch myself again, trying everything I could to try and break something, but I was unable to. The behavior seems to be rock solid, and the PR definitely fixes the issues in `master`.

In the interest of getting this included in WP 5.6, I'm going to go ahead and merge this.
